Add On Device Training Tutorial with ONNX Runtime (#16167)
microsoft/onnxruntime: docs/tutorials/on-device-training.md

---
title: On-Device Training
nav_order: 11
parent: Tutorials
---

# On-Device Training: Building an Android Application

> 

Reviewer: Add an Index to list all the sections to make it easy to navigate.

Reviewer 2: This can be done automatically by the jekyll generator. Have a look at one of the other tutorials to see how this is done.

> 

Reviewer: Add an image/preview of the android app at the top. This will make it easy for the users to see what they are about to build. You can use an existing image may be the last one with Tom Cruise.

For this tutorial, we will leverage the `MobileNetV2` model which has been trained on large-scale image datasets such as ImageNet (which has 1,000 classes). We will use this model for classifying custom data into one of four classes. The initial layers of MobileNetV2 serve as a feature extractor, capturing generic visual features applicable to various tasks, and only the final classifier layer will be trained for the task at hand.

In this tutorial, we will use data to learn to:
- Classify animals into one of four categories using a pre-packed animals dataset.
- Classify celebrities into one of four categories using a custom celebrities dataset.

## Prerequisites

To follow this tutorial, you should have a basic understanding of Android app development using Java or Kotlin. Familiarity with C++ as well as familiarity with machine learning concepts such as neural networks and image classification will help as well.

- Python development environment to prepare the training artifacts
- Android Studio 4.1+
- Android SDK 29+
- Android NDK r21+
- An Android device with a camera in [developer mode](https://developer.android.com/studio/debug/dev-options) with USB debugging enabled

> > **Note** The purpose of this tutorial is to show how ONNX Runtime can be used for On-Device Training. The author of this tutorial is not an expert in Android app development. The information provided in this tutorial is for educational purposes only and should not be considered as professional advice.

Reviewer: This is too formal, not sure if we need this specially the last sentence. @kshama-msft @natke thoughts?

Reviewer 2: Agree with Ashwini on this. Unless you think there is something that they definitely should not use in their own applications

> **Note** The entire android application is also made available on the [`onnxruntime-training-examples`](https://github.com/microsoft/onnxruntime-training-examples/tree/master/on_device_training/mobile/android/c-cpp) GitHub repository.

## Offline Phase - Building the training artifacts

1. Export the model to ONNX.

>     We start with a Pytorch model that has been pre-trained and export it to ONNX. The `MobileNetV2` model has been pretrained on the imagenet dataset that has data in 1000 categories. For our task of image classification, we want to only classify images in 4 classes. So, we change the last layer of the model to output 4 logits instead of 1,000.

Reviewer: nit: We start with a pre-trained PyTorch model and export it to ONNX.

    ```python
   import torch
   import torchvision

   model = torchvision.models.mobilenet_v2(
      weights=torchvision.models.MobileNet_V2_Weights.IMAGENET1K_V2)

   # The original model is trained on imagenet which has 1000 classes.
   # For our image classification scenario, we need to classify among 4 categories.
   # So we need to change the last layer of the model to have 4 outputs.
   model.classifier[1] = torch.nn.Linear(1280, 4)

   # Export the model to ONNX.
   model_name = "mobilenetv2"

>    torch.onnx.export(model, torch.randn(1, 3, 224, 224),

Reviewer 3: A link to documentation on exporting would be useful here. I find this is typically the most difficult step when integrating ONNX Runtime into any existing PyTorch pipeline.

                     f"training_artifacts/{model_name}.onnx",
                     input_names=["input"], output_names=["output"],
                     dynamic_axes={"input": {0: "batch"}, "output": {0: "batch"}})
    ```

2. Define the trainable and non trainable parameters

    ```python
   import onnx

   # Load the onnx model.
   onnx_model = onnx.load(f"training_artifacts/{model_name}.onnx")

>    # Define the parameters that require their gradients to be computed

Reviewer 2: It might be obvious but a verbal connection between requires_grad and trainable parameters could be good here

   # and those that do not.
   requires_grad = ["classifier.1.weight", "classifier.1.bias"]
   frozen_params = [
      param.name
      for param in onnx_model.graph.initializer
      if param.name not in requires_grad
   ]
    ```

3. Generate the training artifacts. We will use the `CrossEntropyLoss` loss and the `AdamW` optimizer for this tutorial.

    ```python

>    from onnxruntime.training import artifacts

Reviewer 3: Is this a new library you've created for LTE? Some documentation on `artifacts` would be helpful.

      onnx_model,
      requires_grad=requires_grad,
      frozen_params=frozen_params,
      loss=artifacts.LossType.CrossEntropyLoss,
      optimizer=artifacts.OptimType.AdamW,
      artifact_directory="training_artifacts"
   )
    ```

4. That's all. The training artifacts have been generated in the `training_artifacts` folder. This marks the end of the offline phase. These artifacts are ready to be deployed to the Android device for training.

## Training Phase - Android application development

1. Set up the project in Android Studio

   a. Open Android Studio and click `New Project`
   ![Android Studio Setup - New Project](../../images/on-device-training-android-studio-setup.png)

   b. Click on `Native C++` -> `Next`. Fill out the `New Project` details as follows:

     - Name - `ORT Personalize`
     - Package Name - `com.example.ortpersonalize`
     - Language - `Kotlin`

      Click `Next`.

   ![Android Studio Setup - Project Name](../../images/on-device-training-android-studio-project-name.png)

   c. Select the `C++17` toolchain -> `Finish`

   ![Android Studio Setup - Project C++ ToolChain](../../images/on-device-training-android-studio-toolchain.png)

>    d. That's it. The Android Studio project has been set up. You should now be able to see the Android Studio editor with some boiler plate code.

Reviewer 3: ```suggestion
   d. That's it! The Android Studio project has been set up. You should now be able to see the Android Studio editor with some boiler plate code.
```

2. Add the ONNX Runtime dependency

   a. Create two new folder called `lib` and `include\onnxruntime` under the cpp directory in the Android Studio project.

   ![lib and include folder](../../images/on-device-training-lib-include.png)

>    b. Head over to [Maven Central](https://central.sonatype.com/artifact/com.microsoft.onnxruntime/onnxruntime-training-android/) and download the `onnxruntime-training-android` archive package (aar file).

Reviewer 3: I went to the page and I didn't see a download package button. Can you add some screenshots on how to download like you had for the Android Studio section?

>    c. Rename the `aar` extension to `zip`. So `onnxruntime-training-android-1.15.0.aar` becomes `onnxruntime-training-android-1.15.0.zip`.

Reviewer 3: Can you just upload a zip to the Maven Central link instead of an .aar file? We can remove this step.

   d. Extract the contents of the zip file.

   e. Copy the `libonnxruntime.so` shared library from the `jni\arm64-v8a` folder to your Android project under the newly created `lib` folder.

   f. Copy the contents of the `headers` folder to the newly created `include\onnxruntime` folder.

   g. In the `native-lib.cpp` file, include the training cxx header file.

   ```cpp
   #include "onnxruntime_training_cxx_api.h"
   ```

>    h. Add `abiFilters` to the `build.gradle (Module)` file so as to select `arm64-v8a`. This setting must be added under `defaultConfig` in `build.gradle`:

Reviewer 3: For this step, maybe including the core for the whole build.gradle module and adding the diff with the ndk lines would look better than just including the new lines. This way people know the code before/after the code you want then to insert.

   ```gradle
   ndk {
      abiFilters 'arm64-v8a'
   }
   ```

>    i. Add the `onnxruntime` shared library to the `CMakeLists.txt` so that `cmake` can find and build against the shated library. To do this, add these lines after the `ortpersonalize` library is added in the `CMakeLists.txt`:

Reviewer 3: Same comment as step h.

>    i. Add the `onnxruntime` shared library to the `CMakeLists.txt` so that `cmake` can find and build against the shated library. To do this, add these lines after the `ortpersonalize` library is added in the `CMakeLists.txt`:

Reviewer 2: shated -> shared

   ```bash
   add_library(onnxruntime SHARED IMPORTED)
   set_target_properties(onnxruntime PROPERTIES IMPORTED_LOCATION ${CMAKE_SOURCE_DIR}/lib/libonnxruntime.so)
   ```

>    j. Let `CMake` know where the ONNX Runtime header files can be found by adding this line right after the above two lines:

Reviewer 3: Merge this instruction with the previous one since its right after. Add the blurb to instruction i.

   ```bash
   target_include_directories(ortpersonalize PRIVATE ${CMAKE_SOURCE_DIR}/include/onnxruntime)
   ```

>    k. Link the Android C++ project against the `onnxruntime` library by adding the `onnxruntime` library to `target_link_libraries`:

Reviewer 3: same comment as step h.

   ```bash
   target_link_libraries( # Specifies the target library.
        ortpersonalize

        # Links the target library to the log library
        # included in the NDK.
        ${log-lib}

        onnxruntime)
   ```

>    l. Build the application and wait for success to confirm that the app has included the ONNX Runtime headers and can link against the shared onnxruntime library successfully.

Author: I think the success of a build will be evident to a developer. I don't think we need to add more screenshots here.

3. Packaging the Prebuilt Training Artifacts and Dataset

   a. Create a new `assets` folder inside the `app` from the left pane of the Android Studio project by right click app -> New -> Folder -> Assets Folder and place it under main.

>    b. Copy the above generated training artifacts to this folder.

Reviewer 3: ```suggestion
   b. Copy the training artifacts generated in step 2 to this folder.
```

>    c. Now, head over to the [`onnxruntime-training-examples`](https://github.com/microsoft/onnxruntime-training-examples/tree/master/on_device_training/mobile/android/c-cpp/data) repo and download the dataset (`images.zip`) to your machine and extract it. This dataset was modified from the orignal [`animals-10`](https://www.kaggle.com/datasets/alessiocorrado99/animals10) dataset available on Kaggle created by [Corrado Alessio](https://www.kaggle.com/alessiocorrado99). To download this dataset to your machine

Reviewer 3: The ending of this instruction seems incomplete. Please review.

Author: Thanks. I think the last sentence is unnecessary.

   d. Copy the downloaded `images` folder to `assets/images` directory in Android Studio.

>    e. The left pane of the project should look like this:

Reviewer 3: ```suggestion
The left pane of the project should look like this:
```

   ![Project Assets](../../images/on-device-training-project-assets.png)

> 4. C++ Code

Reviewer 3: This section contains a lot of code snippets. Would it instead be better to give an english description of the most important functions and then include the specifics of inputs/outputs and implementation details as comments in a cpp file containing all the code? This way we can remove the snippets, give a brief intro, and point people to the cpp file if they're really interested in learning about the specifics of the code.

Author: Particularly about the C++ part: the idea of this tutorial is for people to start writing the code from scratch. They can refer to the cpp file as and when they need since it will be available on github.
We wanted to create a tutorial that started from nothing and built the application.
But I do understand your point about it being too programmatically verbose.

Will leave this unchanged for now.

Author: Removed all the Java code from this tutorial.

                                                            batch_size * image_channels * image_rows * image_cols * sizeof(float),
                                                            input_shape.data(), input_shape.size(),
                                                            ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT));

         std::vector<Ort::Value> output_values;
         output_values.emplace_back(nullptr);

         // get the logits
         session_cache->inference_session->Run(Ort::RunOptions(), input_names.data(), input_values.data(),
                                                input_count, output_names.data(), output_values.data(), output_count);

         float *output = output_values.front().GetTensorMutableData<float>();

         // run softmax and get the probabilities of each class
         std::vector<float> probabilities = Softmax(output, classes.size());
         size_t best_index = std::distance(probabilities.begin(), std::max_element(probabilities.begin(), probabilities.end()));

         return {classes[best_index], probabilities[best_index]};
      }

   } // namespace inference
   ```

>       The classify function invokes another function called `Softmax`. The definition of `Softmax` is:

Author: I think you're right. There is a lot of detail here that can be avoided by asking users to go to the code directly. Not sure hpw others feel @askhade @kshama-msft ?

>       The classify function invokes another function called `Softmax`. The definition of `Softmax` is:

Author: I let the C++ code be here. I removed everything else (the frontend android code).

      Lastly, let's define some utility functions that are needed for the above functions.

   ```cpp
   namespace utils {

      std::string JString2String(JNIEnv *env, jstring jStr) {
         if (!jStr)
               return std::string();

         const jclass stringClass = env->GetObjectClass(jStr);
         const jmethodID getBytes = env->GetMethodID(stringClass, "getBytes",
                                                      "(Ljava/lang/String;)[B");
         const jbyteArray stringJbytes = (jbyteArray) env->CallObjectMethod(jStr, getBytes,
                                                                              env->NewStringUTF(
                                                                                    "UTF-8"));

         size_t length = (size_t) env->GetArrayLength(stringJbytes);
         jbyte *pBytes = env->GetByteArrayElements(stringJbytes, nullptr);

         std::string ret = std::string((char *) pBytes, length);
         env->ReleaseByteArrayElements(stringJbytes, pBytes, JNI_ABORT);

         env->DeleteLocalRef(stringJbytes);
         env->DeleteLocalRef(stringClass);
         return ret;
      }

   } // utils
   ```

> 5. Image Preprocessing in Java or Kotlin

Reviewer 3: Same comment at step 4.

      val height: Int = bitmap.getHeight()
      val stride: Int = width * height

      for (x in 0 until width) {
         for (y in 0 until height) {
               val color: Int = bitmap.getPixel(y, x)
               val index = offset + (x * height + y)

               // Subtract the mean and divide by the standard deviation
               // Values for mean and standard deviation used for
               // the movilenetv2 model.
               buffer.put(index + stride * 0, ((Color.red(color).toFloat() / 255f) - 0.485f) / 0.229f)
               buffer.put(index + stride * 1, ((Color.green(color).toFloat() / 255f) - 0.456f) / 0.224f)
               buffer.put(index + stride * 2, ((Color.blue(color).toFloat() / 255f) - 0.406f) / 0.225f)
         }
      }
   }
   ```

      d. Getting a Bitmap from a Uri

   ```java
   fun bitmapFromUri(uri: Uri, contentResolver: ContentResolver): Bitmap {
      // This function reads the image file at the given uri and decodes it to a bitmap
      val source: ImageDecoder.Source = ImageDecoder.createSource(contentResolver, uri)
      return ImageDecoder.decodeBitmap(source).copy(Bitmap.Config.ARGB_8888, true)
   }
   ```

> 6. Frontend in Java or Kotlin

Reviewer 3: Same comment as step 4.

   c. Copy all the string definitions from [`strings.xml`](https://github.com/microsoft/onnxruntime-training-examples/blob/206521eb22e496b2ea50bef956e63273b6b1d5bf/on_device_training/mobile/android/c-cpp/app/ORTPersonalize/app/src/main/res/values/strings.xml) to `strings.xml` local to your Android Studio.

   d. Copy the contents from [`activity_main.xml`](https://github.com/microsoft/onnxruntime-training-examples/blob/206521eb22e496b2ea50bef956e63273b6b1d5bf/on_device_training/mobile/android/c-cpp/app/ORTPersonalize/app/src/main/res/layout/activity_main.xml) to `activity_main.xml` local to your Android Studio.

   e. Create a new file under the `layout` folder called `dialog.xml`. Copy the contents from [`dialog.xml`](https://github.com/microsoft/onnxruntime-training-examples/blob/206521eb22e496b2ea50bef956e63273b6b1d5bf/on_device_training/mobile/android/c-cpp/app/ORTPersonalize/app/src/main/res/layout/dialog.xml) to the newly created `dialog.xml` local to your Android Studio.

>    f. The remainder of the changes in this secion need to be made in the `MainActivity.kt` file.

Reviewer 4: ```suggestion
   f. The remainder of the changes in this section need to be made in the `MainActivity.kt` file.
```

      val inferButton: Button = findViewById(R.id.infer_button)
      inferButton.setOnClickListener(onInferenceButtonClickedListener)

      val trainButton: Button = findViewById(R.id.train_button)
      trainButton.setOnClickListener(onTrainButtonClickedListener)
      trainButton.isEnabled = false

      val classA: Button = findViewById(R.id.classA)
      classA.setOnClickListener(onClassAClickedListener)
      classA.setOnLongClickListener(onClassALongClickedListener)

      val classB: Button = findViewById(R.id.classB)
      classB.setOnClickListener(onClassBClickedListener)
      classB.setOnLongClickListener(onClassBLongClickedListener)

      val classX: Button = findViewById(R.id.classX)
      classX.setOnClickListener(onClassXClickedListener)
      classX.setOnLongClickListener(onClassXLongClickedListener)

      val classY: Button = findViewById(R.id.classY)
      classY.setOnClickListener(onClassYClickedListener)
      classY.setOnLongClickListener(onClassYLongClickedListener)

      binding.customClassSetting.setOnCheckedChangeListener(onCustomClassSettingChangedListener)

      // Home screen
      binding.statusMessage.text = "ORT Personalize"
   }
   ```

>    j. Custom class button handlers - We would like to use the class buttons for users to select their custom images for training. We need to add the listeners for these buttons to do this. These listeners will do exactly that:

Reviewer 5: looks we almost explain the code in very details? is this needed in this doc, or they are more likely to part of code comments. I know you must already spend lots of time on them, want to know how you guys think about it.

Author: From the discussion we had internally, we wanted to create a very detailed code walkthrough showing how to start a project from start to finish. 
But I feel you're right. I can move all non ort related code out of this file if that is the consensus.

                              optimizerModelPath: String, cacheDirPath: String): Long

   external fun releaseSession(session:Long)

   external fun performInference(session: Long, imageBuffer: FloatArray, batchSize: Int, channels: Int,
                                 frameCols: Int, frameRows: Int, classes: Array<String>): String

   external fun performTraining(session: Long, batches: FloatArray, labels: IntArray, batchSize: Int,
                                 channels: Int, frameRows: Int, frameCols: Int) : Float

   companion object {
      // Used to load the 'ortpersonalize' library on application startup.
      init {
            System.loadLibrary("ortpersonalize")
      }
   }
   ```

   q. One last thing. Add the following in the `AndroidManifest.xml` file to use the camera:

   ```xml
   <uses-permission android:name="android.permission.CAMERA" />
   <uses-feature android:name="android.hardware.camera" />
   ```

7. Running all this on a device.

   a. Let's connect our Android device to the machine and run the application on the device.

   b. Launching the application on the device should look like this:

>    ![Prediction - Cow](../../images/on-device-training-application-landing-page.jpg)

Reviewer 3: This image renders as a super large image, any way to specify the width in markdown?

   d. Run `Training` and wait for the progress dialog to disappear (upon training completion).

   e. Use any animal image from your library for inferencing now.

   ![Prediction - Cow](../../images/on-device-training-application-prediction-cow.jpg)

      As can be seen from the image above, the model correctly predicted `Cow`.

9. Training with a custom dataset - Celebrities

   a. Download images of Tom Cruise, Leonardo DiCaprio, Ryan Reynolds and Brad Pitt from the web.
   
   b. Make sure you launch a fresh session of the app by closing the app and relaunching it.

   c. After the application launches, rename the four classes to `Tom`, `Leo`, `Ryan`, `Brad` respectively using the long click.

   d. Click on the button for each class and select images associated with that celebrity. We can use around 10~15 images per category.

   e. Hit the `Train` button and let the application learn from the data provided.

   f. Once the training is complete, we can hit the `Infer` button and provide an image that the application has not seen yet.

   g. That's it!. Hopefully the application classified the image correctly.

>    ![Prediction - Tom Cruise](../../images/on-device-training-application-prediction-tom.jpg)

Reviewer 3: Same comment as step b.

## Training Phase - Android application development

1. <a name="tp1"></a>Setting up the project in Android Studio

   a. Open Android Studio and click `New Project`
   ![Android Studio Setup - New Project](../../images/on-device-training-android-studio-setup.png)

   b. Click on `Native C++` -> `Next`. Fill out the `New Project` details as follows:

     - Name - `ORT Personalize`
     - Package Name - `com.example.ortpersonalize`
     - Language - `Kotlin`

      Click `Next`.

   ![Android Studio Setup - Project Name](../../images/on-device-training-android-studio-project-name.png)

   c. Select the `C++17` toolchain -> `Finish`

   ![Android Studio Setup - Project C++ ToolChain](../../images/on-device-training-android-studio-toolchain.png)

   d. That's it! The Android Studio project has been set up. You should now be able to see the Android Studio editor with some boiler plate code.

2. <a name="tp2"></a>Adding the ONNX Runtime dependency

   a. Create two new folder called `lib` and `include\onnxruntime` under the cpp directory in the Android Studio project.

>    ![lib and include folder](../../images/on-device-training-lib-include.png)

Author: Added a zoomed in picture.

   d. Extract the contents of the zip file.

   e. Copy the `libonnxruntime.so` shared library from the `jni\arm64-v8a` folder to your Android project under the newly created `lib` folder.

   f. Copy the contents of the `headers` folder to the newly created `include\onnxruntime` folder.

   g. In the `native-lib.cpp` file, include the training cxx header file.

   ```cpp
   #include "onnxruntime_training_cxx_api.h"
   ```

   h. Add `abiFilters` to the `build.gradle (Module)` file so as to select `arm64-v8a`. This setting must be added under `defaultConfig` in `build.gradle`:

   ```gradle
   ndk {
      abiFilters 'arm64-v8a'
   }
   ```

   Note that the `defaultConfig` section of the `build.gradle` file should look like:

   ```diff
   defaultConfig {
      applicationId "com.example.ortpersonalize"
      minSdk 29
      targetSdk 33
      versionCode 1
      versionName "1.0"

      testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
      externalNativeBuild {
         cmake {
               cppFlags '-std=c++17'
         }
      }

   +   ndk {
   +      abiFilters 'arm64-v8a'
   +   }
   

>    }

Reviewer 5: in the page, the ending "}" is in red color, is that expected?

Author: Not expected. I resolved it by adding a leading space.

   a. Let's connect our Android device to the machine and run the application on the device.

   b. Launching the application on the device should look like this:

   <img src="../../images/on-device-training-application-landing-page.jpg"  width="30%" height="30%">

2. <a name="tp8"></a>Training with a pre-loaded dataset - Animals

   a. Let's get started with training using the pre-loaded animals device by launching the application on the device.

   b. Toggle the `Custom classes` switch at the bottom.

   c. The class labels will change to `Dog`, `Cat`, `Elephant` and `Cow`.

   d. Run `Training` and wait for the progress dialog to disappear (upon training completion).

   e. Use any animal image from your library for inferencing now.

   <img src="../../images/on-device-training-application-prediction-cow.jpg"  width="30%" height="30%">

   As can be seen from the image above, the model correctly predicted `Cow`.

3. <a name="tp9"></a>Training with a custom dataset - Celebrities

   a. Download images of Tom Cruise, Leonardo DiCaprio, Ryan Reynolds and Brad Pitt from the web.
   
   b. Make sure you launch a fresh session of the app by closing the app and relaunching it.

>    c. After the application launches, rename the four classes to `Tom`, `Leo`, `Ryan`, `Brad` respectively using the long click.

Reviewer 5: :)
